src/physics/nonlinear_solid.cpp

```diff
@@ -132,49 +148,147 @@ void NonlinearSolid::completeSetup()
     bc.project(displacement_);
   }
 
-  // The abstract mass bilinear form
-  std::unique_ptr<mfem::ParBilinearForm> M_form;
-
-  // The abstract viscosity bilinear form
-  std::unique_ptr<mfem::ParBilinearForm> S_form;
-
   // If dynamic, create the mass and viscosity forms
   if (timestepper_ != serac::TimestepMethod::QuasiStatic) {
     const double              ref_density = 1.0;  // density in the reference configuration
     mfem::ConstantCoefficient rho0(ref_density);
 
-    M_form = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    M = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    M->AddDomainIntegrator(new mfem::VectorMassIntegrator(rho0));
+    M->Assemble(0);
+    M->Finalize(0);
 
-    M_form->AddDomainIntegrator(new mfem::VectorMassIntegrator(rho0));
-    M_form->Assemble(0);
-    M_form->Finalize(0);
+    M_mat.reset(M->ParallelAssemble());
 
-    S_form = displacement_.createOnSpace<mfem::ParBilinearForm>();
-    S_form->AddDomainIntegrator(new mfem::VectorDiffusionIntegrator(*viscosity_));
-    S_form->Assemble(0);
-    S_form->Finalize(0);
+    C = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    C->AddDomainIntegrator(new mfem::VectorDiffusionIntegrator(*viscosity_));
+    C->Assemble(0);
+    C->Finalize(0);
   }
 
-  // Set the MFEM abstract operators for use with the internal MFEM solvers
+  // We are assuming that the ODE is prescribing the
+  // acceleration value for the constrained dofs, so
+  // the residuals for those dofs can be taken to be zero.
+  //
+  // Setting iterative_mode to true ensures that these
+  // prescribed acceleration values are not modified by
+  // the nonlinear solve.
+  nonlin_solver_.nonlinearSolver().iterative_mode = true;
+  nonlin_solver_.SetOperator(residual);
+
   if (timestepper_ == serac::TimestepMethod::QuasiStatic) {
-    nonlin_solver_.nonlinearSolver().iterative_mode = true;
-    nonlinear_oper_ = std::make_unique<NonlinearSolidQuasiStaticOperator>(std::move(H_form), bcs_);
-    nonlin_solver_.SetOperator(*nonlinear_oper_);
+    // the quasistatic case is entirely described by the residual,
+    // there is no ordinary differential equation
+    residual.function = [=](const mfem::Vector& u, mfem::Vector& r) mutable {
+      H->Mult(u, r);  // r := H(u)
+      r.SetSubVector(bcs_.allEssentialDofs(), 0.0);
+    };
+
+    residual.jacobian = [=](const mfem::Vector& u) mutable -> mfem::Operator& {
```

Comment:
At a glance, this lambda and the remaining lambdas can also capture a reference to the calling object + not need `mutable`.